gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -109,12 +132,23 @@ def self.signature_for_method(method)
   # Only public because it needs to get called below inside the replace_method blocks below.
   def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
     current_declaration = T::Private::DeclState.current.active_declaration
-    return if !current_declaration
-    T::Private::DeclState.current.reset!
-
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
     original_method = mod.instance_method(method_name)
 
+    if final_method?(original_method)
+      raise "`#{mod.name}##{method_name}` was declared as final and cannot be redefined"
+    end
+    mod.ancestors.each do |a|
+      (a.instance_methods(false) + a.private_instance_methods(false)).each do |m|
+        if m == method_name && final_method?(a.instance_method(method_name))
+          raise "`#{a.name}##{method_name}` was declared as final and cannot be overridden in `#{mod.name}`"
```

Comment:
```suggestion
          raise "`#{a.name}##{method_name}` was declared as final in `#{mod.name}` and cannot be overridden"
```

Reply:
I don't think this is necessarily true. The method was declared as final in `a`, and redeclared (possibly with a `final` sig, possibly not) in `mod`.